PIG 13 - Gammapy dependencies and distribution

cdeil opened the pull request:
This PIG is to make a clear plan and gather feedback concerning Gammapy dependencies and distribution.

Please give me a day to work on this, don't comment yet.

bsipocz commented on docs/development/pigs/pig-011.rst:
```diff
@@ -0,0 +1,138 @@
+
+Actually, there is one platform where binary wheels for Gammapy and all
+dependencies exits already. If you ``pip install gammapy`` on Raspberry Pi,
+you'll get binary wheels thanks to this guy:
```
is it only 1 person or rather 2?

cdeil replied:
I'll rewrite the wheels section completely now.

bsipocz commented on docs/development/pigs/pig-011.rst:
```diff
@@ -0,0 +1,138 @@
+Gammapy with pip. If you use Macports and want ``port install py37-gammapy``,
+let us know!
+
+Others
```
Does Ole package gammapy for Debian? I believe most/all affiliated packages are included?

cdeil replied:
I'll clarify Debian and other distros for Gammapy, See #324

cdeil commented on docs/development/pigs/pig-011.rst:
```diff
@@ -0,0 +1,138 @@
+isn't very high: if you want or need it, let us know!
+
+Sunpy has already dropped Python 3.5 support in v1.0. Astropy still supports
+Python 3.5, they will move to 3.6+ with Astropy 4.0 in early 2010.
```
Fixed

cdeil commented on docs/development/pigs/pig-011.rst:
```diff
@@ -0,0 +1,138 @@
+Abstract
+========
+
+As preparation for the planned Gammapy v1.0 release this fall, we propose to
```
fixed

cdeil commented on docs/development/pigs/pig-011.rst:
```diff
@@ -0,0 +1,138 @@
+
+- We're not aware of any users that have to or even want to use Python 3.5, most
+people use Gammapy on Python 3.6 or 3.7 now via conda.
+- Reduce maintenance effort. conda-forge now only builds binaries for Python 3.6
```
We have e.g. regions as a dependency that has C extension and doesn't have binary wheels, and soon astropy-healpix. So at the moment using `pip` to install Gammapy dependencies is not great, it will run a few minutes to compile things. (might just be 1-2 minutes).

cdeil commented:
@bsipocz - thanks for the review. This is super early, the structure and text of the PIG isn't good yet. I'll ping you here again for review once it's in a shape where it makes sense.

There's one key question though that we should discuss this early already: @adonath and @registerrier - please have a look [here](https://github.com/gammapy/gammapy/blob/420000d48758a7cb6da39d82344ca4a74a5ba26d/docs/development/pigs/pig-011.rst#summary) and comment concerning which `regions` version we support. If we put `region.rotate` in now (see https://github.com/astropy/regions/issues/265), it will be in regions v0.5 from June or July 2019 only. And if we want to clean up the rotated regions code in Gammapy before Gammapy v1.0, and not spend much time e.g. to bundle or monkey-path parts or regions or such horrible things, we'll depend on a super-recent version of regions. We'll have the same for astropy-healpix, where 1-2 small things are missing so that we can use it.

If we have these two Astropy-affiliated packages as dependencies, with versions from summer 2019, then the motivation to support very old version of Numpy, Scipy, Astropy is much reduced. The simplest option for us would be to require fairly modern versions there then, that are e.g. form 2017 or 2018. I'm in favour of this, because most of our users use conda, and there we just have a nice way to ship recent software on all platforms.

On the other hand, I don't think there really are things in the latest Numpy, Scipy, Astropy that we would take advantage of in Gammapy at this point - requiring modern versions there would mostly mean less effort on testing for old versions.

The other major system where Gammapy and the dependencies are packaged is Debian.
@olebole - could you please also have a look [here](https://github.com/gammapy/gammapy/blob/420000d48758a7cb6da39d82344ca4a74a5ba26d/docs/development/pigs/pig-011.rst#summary) and let us know what you recommend so that packaging Gammapy remains possible in Debian, i.e. we're not too agressive with new versions?

Our main goal is to have a good Gammapy v1.0 in Oct 2019 (see our [roadmap](https://docs.gammapy.org/0.12/development/pigs/pig-005.html#releases)), but I'd like to set up the dependencies and testing already now in the coming months to match what we plan to do for v1.0.

Thoughts?

bsipocz commented:
> If we have these two Astropy-affiliated packages as dependencies, with versions from summer 2019, then the motivation to support very old version of Numpy, Scipy, Astropy is much reduced. The simplest option for us would be to require fairly modern versions there then, that are e.g. form 2017 or 2018. I'm in favour of this, because most of our users use conda, and there we just have a nice way to ship recent software on all platforms.

it can be pointed out in the docs of not supporting doesn't necessarily mean that old versions are broken with gammapy, rather that that they are not tested and regressions affecting only those old releases will most probably won't be fixed.

bsipocz commented:
(that would help to reduce the CI matrix significantly,  currently travis totalling to ~3 hours is way to long)

olebole commented:
> The other major system where Gammapy and the dependencies are packaged is Debian.
@olebole - could you please also have a look here and let us know what you recommend so that packaging Gammapy remains possible in Debian, i.e. we're not too agressive with new versions?

In principle, we try to keep up-to-date with our testing/unstable releases (with the exception of the current "freeze", where we don't update versions). The versions in the link look therefore completely OK.

Two general (minor) points here, however: First, a bug that appeared in Ubuntu (astropy/astropy#8783) made me think about being a bit more conservative (with no solution yet; input is welcome).

Second, apart from Ubuntu (which comes out 6monthly), the next (post-Buster) Debian release will be out in more than two years only, and Debian Buster will carry gammapy 0.10 until then. In principle, it can be "backported" (made available in a special Backports add-on repository), but this requires to have the dependencies available, so they may need to be backported as well. Hard dependencies on the newest, shiniest versions everywhere makes this harder (up to impossible when there are conflicts f.e. in the Python version). So, I can only give the general recommendation to keep the dependency versions a bit relaxed if possible, especially when they require big updates in turn. However, handling backports is still not systematic anyway (volunteers ... where one can find volunteers...?)

cdeil commented:
@olebole and @bsipocz - Thanks for the feedback!

Concerning Debian and Linux distributions: my feeling is that it's not worth to spend time on backports. We don't have anyone familiar with Linux packaging in the Gammapy team, can't contribute there at the moment. And Gammapy will change a ton in the next half year until v1.0 at least, with releases every 2 months. We don't have the time to support old releases, we're trying to go as fast as we can towards a good v1.0 in October 2019. That's just not appropriate for Linux distributions which have ~ yearly update cycles. After v1.0 and in the coming years the story might be very different, as Gammapy becomes stable having Debian packages becomes very valuable.

One thing that changed just now is that Tom set up https://github.com/astropy/astropy-wheels so that binary wheels for all Gammapy dependencies can be distributed. If we can get in on that service, I think we should try it for a bit and distribute binary wheels as well. That's a second means besides conda to distribute quickly on all platforms, and it should work well on top of e.g. a system Python on Debian for now, while our code changes too rapidly for the system packaging there.
 
I will update this PIG in the coming days to a useful draft state, then ask for review again.

One more thing: I noticed that PIG 11 was already given the the lightcurve PIG in #1451 .
So I'll change this PIG on Gammapy dependencies and distribution now to be PIG 13.

adonath commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,254 @@
+pyyaml           unclear          3.13 (Jul 2018)
+click            unclear          7.0  (Sep 2018)
+===============  ===============  ===============
+
```
Are there LTS releases of the packages, we should take care of? Otherwise I'm fine with the selection...

bsipocz replied:
I don't think that you should support lts 2.0 astropy going forward, as the main reason for that is to support python2. I'm  then unaware any of the other packages do real LTS releases  (except some extended  support for their  last python2 supporting versions).

adonath replied:
Thanks @bsipocz for the clarification.

cdeil replied:
I didn't add any comments concerning LTS. I don't think anything is needed, but @adonath if you think something should be added, let me know. Concerning Gammapy itself, I can add a sentence at the end in the outlook of the PIG that post-v1.0 releases and the question whether we do LTS and backports is not addressed here, that's a separate PIG.

adonath commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,254 @@
+
+Optional dependencies
+=====================
+
```
May be put a short definition of "optional dependency" here.

cdeil replied:
will do

adonath commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,254 @@
+
+Required dependencies
+=====================
+
```
Maybe but a short definition of "required dependency" here.

cdeil replied:
Done  - but probably can be improved further. Give me a day, and then please read the PIG and comment again.

adonath commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,254 @@
+pyyaml           unclear          3.13 (Jul 2018)
+click            unclear          7.0  (Sep 2018)
+===============  ===============  ===============
+
```
Maybe add  a short description of the required dependencies and what they are used for in Gammapy. Also outline the reason for the choice and why they are "required". The reasoning for choosing Astropy, Numpy and Scipy is easy, but for pyyaml and click it's less clear, because there are alternatives...

cdeil replied:
done

bsipocz commented on docs/development/pigs/pig-013.rst:
```diff
@@ -15,19 +15,25 @@ PIG 13 - Gammapy dependencies and distribution
+
+Given that we have a good way to ship recent versions of Gammapy and all
+dependencies to users on all platforms, we suggest to drop support for Python
+3.5 or later, and to increase the minimum supported version of our dependencies.
```
> we suggest to drop support for Python
3.5 or later, 

This reads a little bit ambiguous, do you mean to say to go for 3.5+ or 3.6+ or 3.7+ here?

bsipocz commented on docs/development/pigs/pig-013.rst:
```diff
@@ -15,19 +15,25 @@ PIG 13 - Gammapy dependencies and distribution
+supported distribution channels, to be introduced in Gammapy v0.14, and then to
+be used in Gammapy v1.0 and likely without much changes moving forward.
+
+Given that we have a good way to ship recent versions of Gammapy and all
```
I suggest a bit of rephrasing. Gammapy is probably OK in this sense as it's still very rapidly developing, but the typical reasons for not going with latest and greatest is not that it's difficult to upgrade, but to avoid any potential upgrading issues with pipeline code (e.g upgrading gammapy might be OK, but if it also means a need to update numpy then observatories/pipelines become worried).

cdeil replied:
I rephrased and tried to clarify a bit. But the abstract is already too long, didn't want to add much here. Gammapy isn't used in pipelines yet, and e.g. proper support for CTA pipelines and version relation to ctapipe and Gammapy is beyond the scope of this PIG. For now, both are in the prototyping phase, and ctapipe is OK to always run on the latest and greatest for now, i.e. with Gammapy we're mostly catching up here.

bsipocz commented on docs/development/pigs/pig-013.rst:
```diff
@@ -15,19 +15,25 @@ PIG 13 - Gammapy dependencies and distribution
+Concerning distribution, we suggest to try to build and distribute Python binary
+wheels to get a fast ``pip install gammapy``, and if it works well and the extra
+maintenance effor is low, to support that distribution channel for Gammapy v1.0
+and moving forward. We suggest to remove support for distribution via Macports.
```
> We suggest to remove support for distribution via Macports.

I would suggest to change the sentence to leave the option open for 3rd party packagers to package up gammapy, e.g. it'll stay in debian because they support it, not because gammapy devs support it --> "Gammapy developers do not actively support any other distribution channels."

cdeil replied:
Yes, important point. I put a new sentence.

bsipocz commented on docs/development/pigs/pig-013.rst:
```diff
@@ -37,85 +43,128 @@ lot. Thanks to conda, we now have a way to ship up-to-date versions of Gammapy
+  standard and better support for Time columns in Table, the new
+  ``astropy.uncertainty`` package which we could use (see :ref:`pig-14`_), the
+  ``SkyCoord.directional_offset_by`` method for field of view coordinates,
+  common WCS API, the new ``astropy.timeseries``, ASDF, ...
```
`...`--> `etc.` or even better start the listing with  `for example`

cdeil commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,254 @@
+============
+
+In recent years, the Python packaging and distribution system has improved a
+lot. Thanks to conda, we now have a way to ship up-to-date versions of Gammapy
```
done

cdeil commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,254 @@
+
+Notes:
+
+- We already mentioned the possibility to drop Python 3.5 support in
```
done

cdeil commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,254 @@
+v0.13 release in late July 2019, which will ship to users in Gammapy v0.14 in
+September 2019.
+
+As mentioned already in :ref:`pig-013`, there is always a benefit and a cost to
```
yes, done.

cdeil commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,254 @@
+Given this situation, we feel that it is OK to drop support for old versions
+(most notably Python 2.7, see :ref:`pig-013`), and to change a few Gammapy
+dependencies from "optional" to "required". The difference is that a required
+dependency is in the ``install_requires`` list in ``setup.py``, which means that
```
I introduced sub-sections in the introduction, and added a section at the start of "required" and "optional" sections. Should be better now, but please read and comment again tomorrow - for sure the PIG can be improved further.

cdeil commented on docs/development/pigs/pig-011.rst:
```diff
@@ -0,0 +1,138 @@
+
+As preparation for the planned Gammapy v1.0 release this fall, we propose to
+clarify which Gammapy dependencies are required and which ones are optional. We
+also propose to increase the minimum required version for some dependencies.
```
We don't have LTS releases or policies like "we support last X releases or Y years of Numpy". At the moment Gammapy is in the prototyping phase, and our goal is to ship Gammapy and it's dependencies to users only.

cdeil commented on docs/development/pigs/pig-011.rst:
```diff
@@ -0,0 +1,138 @@
+- Allow the use of new language features, specifically f-strings and dict
+everywhere (instead of a mix of OrderedDict and dict)
+
+The first point is the most important one: the cost of supporting Python 3.5
```
I think this is addressed. The PIG clearly proposes to drop Python 3.5 support, I don't expect this to be controversial, the vast majority of our users are on conda and there 3.5 support is already dropped.

cdeil commented on docs/development/pigs/pig-011.rst:
```diff
@@ -0,0 +1,138 @@
+Sunpy has already dropped Python 3.5 support in v1.0. Astropy still supports
+Python 3.5, they will move to 3.6+ with Astropy 4.0 in early 2010.
+
+tbd: decide on min version for other dependencies
```
@adonath @registerrier and I had a call, and decided to be very recent in the min versions we support, to make our lives easier and to be able to go faster with dev in the coming months. And only if we get backlash on this in the PIG commenting period, we'll backpaddle.

cdeil commented:
@adonath @bsipocz - I think I've addressed all of your comments in 30cdf2c47 .

Please give me a day or two to finish the PIG before reading and commenting again. I'll post a comment here when it's ready for review by the co-authors, hopefully already tomorrow.

bsipocz commented on docs/development/pigs/pig-013.rst:
```diff
@@ -239,6 +239,10 @@ However, we suggest to stick to ``gammapy`` as a package with a very small set
 of dependencies, and then use the environment specifications to create the stack
 and execution environment that most users want, not metapackages for now.
 
+TODO: now I'm having second thoughts. Maybe we should go the pip & conda metapackages route?
```
doesn't @astrofrog have similar setup for glue? Maybe experiences from there could help with the decision for gammapy.

cdeil commented:
@Bultako @jasercion @bsipocz @astrofrog or @mackaiver (xref #1245) or anyone - what do you think is the best way to ship both a core package with required dependencies, as well as an analysis stack that includes optional dependencies, with both pip and conda?

If you have time to read [this draft proposal section for Gammapy](https://github.com/gammapy/gammapy/blob/b1c4374f0d2572c22584fb0c7a35489d0b20d0de/docs/development/pigs/pig-013.rst#environments) and leave a comment here, that would be super helpful!

cdeil commented:
Advice from Tom Robitaille via Astropy Slack:

> For pip, I prefer `extras_require` if it's for actual optional dependencies to a single package, and a metapackage on PyPI if you want to install multiple different gammapy components. For conda, I think a meta-package is probably the best way if you don't want to install all the optional dependencies with the default build. In my experience, requirements.txt (for pip) and environment files (for conda) will only be usable by a small fraction of users.

> Just to expand on the metapackage vs extras_require difference for pip, I'd say that if it's that e.g. MPL or scipy are optional dependencies, that should be an extras_requires. But for e.g. glue, I have a glueviz meta-package that installs the core glue package as well as any plugins I want to recommend to users.

mackaiver commented:
Just pinging @maxnoe. I'm not an expert in python packaging.

MaxNoe commented:
So from my understanding, packages should not use `requirements.txt`, this is something for applications or just some random scripts that have no `setup.py`.

For packages, dependencies should be put in the `install_requires`, `setup_requires` etc. arguments of the `setup` function.

cdeil commented:
@adonath @registerrier @bsipocz as co-authors, or anyone that has time now: Please review the first complete draft of the PIG [here](https://github.com/gammapy/gammapy/blob/4d0594182a54450bfd61ac9381448c60f0673efb/docs/development/pigs/pig-013.rst).

If possible, I would like to send this for review to the Gammapy team and users soon, but of course before that we should agree on the proposal, so please focus comments on the main questions.

One change I made, based on the feedback from Tom Robitaille and @MaxNoe above is to not propose to create and distribute `requirements.txt` files to users via `gammapy.org`, but instead to define an "all" target and have `pip install gammapy[all]`. I've also made it clear that this will be an experimental way to distribute Gammapy, it's not clear if we have time to work on Wheels for now, or how well those will work. So I mention it as something we want to work on, but it's not a requirement or blocker for Gammapy v1.0.

For conda there's the same question. For now I'm suggesting we stick to what we have with the env file, because it's been working well for us and as far as I know users like it. However, I do mention the option to change to or add a conda meta-package called `gammapy-all` as something we invite comments on. If someone is interested to create that and maintain it, we could also just try it out. To me it's not clear how maintenance / versioning would work exactly, e.g. in the current env files we sometimes found little issues post release and just updated the dependency versions there via "stealth updates", i.e. without a patch version number like "0.12.1"; is this possible with conda meta-packages our would we have to tag bugfix versions there?

The core of the proposal remains for dependencies: we clearly list which ones we have, and propose adopting recent versions, which will help us clean up the code and focus more on features than on maintenance in the coming months.

registerrier commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,451 @@
+- Change ``astropy-healpix`` to the CDS implementation, release a stable version and
+  check conda and pip binaries (`astropy-healpix GH 128`_)
+- Change Gammapy from ``healpy`` to ``astropy-healpix`` (`GH 1167`_)
+- Maybe: rewrite spectral model error evaluation code to not use the
```
Maybe leave that for PIG14

cdeil replied:
OK, removed that point here. (although there's no problem in having some overlap in PIGs, the important thing is to get good solutions, agree on them, and implement them, quickly)

registerrier commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,451 @@
+  new CI test matrix, but this could also come later.
+- Update continuous integration test matrix to test against the minimum required
+  versions as specified here (and also newer versions in addition) (`GH 2270`_).
+  Maybe move completely from travis-ci to Azure?
```
Is it related?

cdeil replied:
I think the CI update should be mentioned. It's one of the larger pieces of work actually, and is absolutely needed to ensure that what we describe here works and is supported. I removed the ? concerning travis-ci vs azure though, that's something we'll just decide when we work on this.

registerrier commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,451 @@
+- Review, restructure and update all Gammapy installation instructions
+- Change ``astropy-healpix`` to the CDS implementation, release a stable version and
+  check conda and pip binaries (`astropy-healpix GH 128`_)
+- Change Gammapy from ``healpy`` to ``astropy-healpix`` (`GH 1167`_)
```
Maybe those two points should be merged into one.

cdeil replied:
I prefer to keep those two points separate. First we have to work on `astropy-healpix` and release a new version. Then we can change Gammapy to use it.

MaxNoe commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,451 @@
+distribution channel, used in our testing continuous integration setup) only
+contains Python 2.7, 3.6 and 3.7 at this point (and 3.8 added in fall 2018),
+i.e. testing on Python 3.5 is already extra effort. Also, Python 3.6 allows us
+to use f-strings and just dict instead of a mix of dict and OrderedDict like we
```
Also, ordern only guaranteed since python 3.7, in CPython 3.6 it is only an implementation detail

cdeil replied:
Changed to "Also, Python 3.6 contains some nice new features that developers can use."

MaxNoe commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,451 @@
+
+The ``pyyaml`` package is the most-used YAML package for Python, used e.g. also
+in Astropy core for ECSV or ASDF I/O. The ``click`` package was chosen to
+implement the ``gammapy`` command line interface. For ``pyyaml`` and ``click``
```
I would strongly recommend against pyyaml. It uses an outdated spec (1.1) of yaml and will not be updated to support a newer one. 

Conda uses ruamel_yaml, which supports yaml 1.2.

I think this especially important for astropy/gammapy as only yaml 1.2 supports scientific notation floats like `1.5e3`

bsipocz replied:
@MaxNoe -  could you also raise this as an issue on the astropy issue tracker?

cdeil replied:
@maxNoe - Thank you for the suggestion! That is exactly the kind of discussion and feedback I had hoped to trigger via this PIG.

I'm not sure if `pyyaml` or `ruamel_yaml` is the better choice, still suggest to stick with `pyyaml` for now.

Yes, `ruamel_yaml` has some nice things, and it's weird that pyyaml doesn't parse `1.5e3` as a float, but requires one to write `1.5e+3` instead. The round-tripping feature including comments is for me the biggest improvement over `pyyaml`, that's really nice if we want to read and write back modified YAML files ever in a way that is as unintrusive as possible (not clear if we need that - maybe for config files).

On the other hand, with `ruamel_yaml` I have concerns with stability and support. E.g. https://pypi.org/project/ruamel.yaml/ says **For production systems you should pin the version being used with ``ruamel.yaml<=0.15** and it's not really clear to me if / what they plan to change in v0.16 or v1.0 for `ruamel_yaml`, and if that ever will come or not, i.e. if and when things would break in Gammapy. It looks like it's mostly one guy developing `ruamel_yaml` on Bitbucket and he has ~ 100 packages with `ruamel` in the name, and I'm not sure what that is?

Astropy and ASDF use `pyyaml`, in Astropy for ECSV and other things. It seems stable and well-supported, there's conda binaries and pip wheels and I never had an install issue over the past years.

So if we adopt `ruamel_yaml` in Gammapy, we effectively depend on both, in addition to `pyyaml`, kind of doubling the chance of trouble for us in the future (e.g. API changes, or packaging screw-ups, ...). For that reason I would suggest we stick with `pyyaml` for now, and only mention the situation and `ruamel_yaml` in the alternatives section, as something we might change to later if Astropy and the larger Python community adopts it over `pyyaml` (which might or might not happen).

MaxNoe replied:
0.16 is already out, the changed the API from 

```
from ruamel.yaml import safe_load # or load
config = safe_load(open('config.yaml'))
```

to a classed based loader, that takes configuration like this:

```
from ruamel.yaml import YAML

yaml = YAML(typ='safe')
config = yaml.load(open('config.yaml')
```

As mentionend, conda itself already depends on ruamel yaml, so it's basically guarenteed to be available with conda and no problem to get it with pip.

MaxNoe replied:
The changes announced are the removal of the old api, which is still supported but deprecated in 0.15.

cdeil replied:
I've split pyyaml and click into their own paragraphs, it wasn't good to mix them into one.

@registerrier - for click I've said it will be re-considered in the high-level PIG. That one will probably take longer, shouldn't hold up this one. We can implement this and then still change our mind in the coming months, that's fine. The key proposal here is to drop support for the old versions, and because that might be controversial I went for a PIG.

@maxnoe - for YAML I have a preference to stick with `pyyaml`, sorry.
I've added a longish section trying to summarise the pros and cons of `pyyaml` and `ruamel_yaml` and stated why we prefer to stick with `pyyaml` for now.
See 0080a4bc6 .
This PIG will go to official review phase soon, and we might re-discuss and re-consider this, but for now I'm marking this discussion as resolved.

cdeil commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,451 @@
+Python binary wheels for Gammapy, regions, reproject and astropy-healpix for
+Linux, macOS and Windows, so that users don't need a compiler and installation
+is fast. We suggest to define a list of dependencies "all" in ``extras_require``
+via setuptools, so that ``pip install gammapy[all]`` is mostly equivalent to
```
I put "so that ``pip install gammapy[all]`` installs the optional Gammapy dependencies."

cdeil commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,451 @@
+
+In recent years, the Python packaging and distribution system has improved a
+lot. Thanks to conda, we now have a way to ship up-to-date versions of Gammapy
+and all dependencies on all supported platforms, the number of installation
```
done

cdeil commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,451 @@
+In recent years, the Python packaging and distribution system has improved a
+lot. Thanks to conda, we now have a way to ship up-to-date versions of Gammapy
+and all dependencies on all supported platforms, the number of installation
+issues reported by users has been very low in recent years.
```
Added "in a dedicated environment for each version"

cdeil commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,451 @@
+somewhere.
+
+Currently we have both ``pip install gammapy`` and ``conda install gammapy``
+only install the required dependencies. We suggest to keep it like that.
```
done

cdeil commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,451 @@
+Tensorflow continues to mature and improve in the coming years (see ``jax`` and
+``tensorflow-probability``), that could be considered for Gammapy.
+
+It is not described here, but very likely after implementing this PIG we will
```
I think this is relevant, in the sense that it shows that we have a plan how to test and support what is described in this PIG. I've changed to this paragraph to try and explain better why we mention it:

After implementing this PIG we will follow up by improving Gammapy testing (`GH
2270`_) and by modernising the Gammapy packaging following the recommendations
in `APE 17`_. We plan to improve the test matrix to ensure that all supported
versions of dependencies work, and also that distribution works, by packaging
Gammapy and installing it with pip and conda and then executing the tests. This
is not considered controversial and thus not described in detail in this PIG.

cdeil commented:
I've added a section proposing we stick with Cython for now, and only re-consider using Numba later.

cdeil commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,451 @@
+===============  ================  ================
+Dependency       Gammapy 0.12      Gammapy 0.14
+===============  ================  ================
+ipython=7.5      7.3   (Feb 2019)  7.6   (Jun 2019)
```
copy & paste error from the env YAML file. Fixed.

cdeil commented:
I've updated the PIG. I think it is in a state where we could go for official review, e.g. with a deadline of July 17 (one week if we circulate tomorrow).

@adonath @registerrier @bsipocz as co-authors or anyone - thoughts?
Are you happy with this proposal? Or would you prefer we do something differently?

cdeil commented:
I realised that we're not saying yet if we'd put `=` or `>=` for the required and optional dependencies in `setup.py`, or no version constraint for some things like `click` which basically have worked with any version in the past years, i.e. the version given in the PIG would only be what we test, but we wouldn't require it to install Gammapy.

I think it's very uncommon to put `=` in the package requirements, so suggest to put `>=` for most or all required and optional dependencies, never `=`. That's e.g. what Sunpy and most other Python packages do, see e.g. [here](https://github.com/sunpy/sunpy/blob/73eb7ec9d7f082f2c33dd601852dd54051b2ab82/setup.cfg#L28-L65).

Of course that means that install of old Gammapy versions will likely break at some point in the future, e.g. if in 3 years matplotlib 4.0 comes out and has a backwards incompatible change and breaks our use of it, then `pip install gammapy=1.0` would install matplotlib 4.0 then if we specify `matplotlib>=3.0` now.

So `install_requires` and `extras_require` would not give a reproducible execution environment. For that, something extra is needed. If we really want to ship this via pip, we'd have to make a `gammapy-all` metapackage or go back to the `requirements.txt` idea.

Not sure what to put in the PIG for now, we have to decide and clarify this point somehow. Kind of wishing more and more I hadn't added the distribution and pip aspect and just limited the PIG to the proposal to require Python 3.6 and very modern dependencies.

Thoughts?

cdeil commented:
@adonath @registerrier and @bsipocz as co-authors, or anyone interested: please do a final review of this PIG. I didn't manage to finish this up before the vacation break, but now I think it's in decent shape and would like to circulate this with a deadline for comments of ~ 10 days.

Note that I significantly simplified and shortened the proposal. Basically only the core suggestion to update to modern dependencies remains, and quite a bit of explanation concerning dependencies and distribution, but without actually proposing and asking for approval on any changes there.

Personally I found other considerations and discussions we had here and offline e.g. on pip and wheels, or on Cython vs Numba etc. useful, but I don't see value on elaborating on those much in the PIG.

adonath commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,271 @@
+question how to ship an environment. This could be prototyped any time, in
+parallel with the existing solution, to gain familiarity with metapackages.
+
+We could support other ways to ship Gammapy, e.g. Homebrew (works on MacOS and
```
I think this rather belongs to the outlooks section. It's not really an alternative, but rather optional.

cdeil replied:
done

adonath commented on docs/development/pigs/pig-013.rst:
```diff
@@ -0,0 +1,271 @@
+We could support other ways to ship Gammapy, e.g. Homebrew (works on MacOS and
+also Linux now), or Docker images (works anywhere, self-contained).
+
+We note that we expect the Gammapy dependencies to change from time to time.
```
The same here, I would move it the "outlook" section. It's not an alternative to the proposed changes.

MaxNoe replied:
Is that really noteworthy though? I would just drop it. Updating and replacing dependencies is at the very core of every software developement.

cdeil replied:
Removed.

cdeil commented:
@adonath @MaxNoe  - Addressed your comments in 153606125 .

adonath commented:
@cdeil No more comments from my side...

bsipocz commented:
@cdeil - I'm travelling on a workshop and then will travel next week, so it's not realistic to promise any reasonable review before I'm back to Seattle.

cdeil commented:
Comment period for this PIG ended yesterday: https://groups.google.com/forum/#!topic/gammapy/B7Cj761enN4
We didn't get any further comments.
I'll merge this in now, open reminder issues for the tasks for myself and implement them within the next week.

adonath commented:
🎉

bsipocz commented:
@cdeil - this is now water under the bridge, but skimming through the PIG again, I started to wonder what will happen in the future? E.g. this PIG doesn't set up rules for future dependency version drops. Is the plan to set up a table for each new (major) version, or rather define a set of rules?

cdeil commented:
The idea is to transfer the tables to the docs, and then to maintain them there (likely making incremental changes in pull requests without much discussion).

We don't have a policy or something like that for the coming years yet, that will come later.

This proposal was just to have an opportunity for discussion and feedback and be transparent about this one-time change.

bsipocz commented:
OK, that sounds good.
